lib/ecto_mnesia/adapter.ex

```diff
@@ -43,65 +43,70 @@ defmodule Ecto.Mnesia.Adapter do
   """
   def prepare(operation, %Ecto.Query{from: {table, schema}, order_bys: order_bys, limit: limit} = query) do
     ordering_fn = order_bys |> Ordering.get_ordering_fn()
-    context = table |> Context.new(schema)
     limit = limit |> get_limit()
-    {:nocache, {operation, query, limit, context, ordering_fn}}
+    limit_fn = if limit == nil, do: &(&1), else: &Enum.take(&1, limit)
+    context = table |> Context.new(schema)
+    {:nocache, {operation, query, {limit, limit_fn}, context, ordering_fn}}
   end
 
   @doc """
   Perform `mnesia:select` on prepared query and convert the results to Ecto Schema.
   """
```

Comment:
Found a TODO tag in a comment: # TODO: deal with `opts`: `on_conflict` and `returning`